### PR TITLE
Merged QuakeWorld qc Code into codebase.

### PR DESCRIPTION
**Pull request by Zungrysoft**
Merged the NetQuake and QuakeWorld codebases into a single codebase so any changes we make will support both engine types. We will now be able to distribute a qwprogs.dat for QW deathmatch play.

You can compile the codebase for either NetQuake (creating progs.dat) or QuakeWorld (creating qwprogs.dat). You can compile for QuakeWorld by passing the argument `-D__QW__` into fteqcc. Otherwise, it will compile for NetQuake.

Wherever there is a discrepancy between the NQ and QW implementations, here is how I resolved it:
- If it only affects singleplayer/coop, use the NQ implementation.
- If it only affects deathmatch, use the QW implementation.
- If it affects both:
	- If the difference is required due to differences in communication with the engine or available engine features, branch the code using preprocessor directives.
	- Otherwise, branch the code based on whether deathmatch is currently enabled.

Or in short, singleplayer behaves the same as it did before, but deathmatch now always behaves like QuakeWorld, regardless of which engine type is being used.

QW Source used for this merge: https://github.com/id-Software/Quake/tree/master/qw-qc

I'm still working on testing this, though it's ready for review.

### CODE REVIEW

File: build_components.json
Comment (MotoLegacy) [inline build_components.json:27]: Why was the pak ordering here changed?

File: build_components.json
Comment (MotoLegacy) [inline build_components.json:32]: Does vanilla QuakeWorld support loading qwprogs.dat externally?

File: qcsrc/deathmatch_settings.qc
Comment (MotoLegacy) [inline qcsrc/deathmatch_settings.qc:2]: For most of the functions in here, we can use #define macros instead since they are returning comparison values. This will keep our progs a little smaller in size and provide some slight speedups which may prove beneficial for very low end machines.

File: qcsrc/defs.qc
Comment (MotoLegacy) [inline qcsrc/defs.qc:409]: Channel 0 actually is used, I think it serves as a fine fallback but you should revise this comment

File: qcsrc/defs.qc
Comment (MotoLegacy) [inline qcsrc/defs.qc:455]: Nitpick: #define these maybe

File: qcsrc/doors.qc
Comment (MotoLegacy) [inline qcsrc/doors.qc:66]: What? Does QW see this as a valid channel?

File: qcsrc/localized_text.qc
Comment (MotoLegacy) [inline qcsrc/localized_text.qc:68]: Was changing the indentation here your doing or the editors?

File: qcsrc/localized_text.qc
Comment (MotoLegacy) [inline qcsrc/localized_text.qc:274]: What is the justification for changing this? Just curious

File: qcsrc/localized_text.qc
Comment (MotoLegacy) [inline qcsrc/localized_text.qc:339]: I like this, good one

File: qcsrc/misc.qc
Comment (MotoLegacy) [inline qcsrc/misc.qc:247]: Is this a fix? Is it supposed to be 20?

File: qcsrc/server.qc
Comment (MotoLegacy) [inline qcsrc/server.qc:1]: Can you rename this to better denote that it's QW-specific?

File: qcsrc/spectate.qc
Comment (MotoLegacy) [inline qcsrc/spectate.qc:6]: "PURPOSE"?

File: qcsrc/world.qc
Comment (MotoLegacy) [inline qcsrc/world.qc:260]: Hm?

File: qcsrc/world.qc
Comment (MotoLegacy) [inline qcsrc/world.qc:367]: Are these cvars defined in vanilla Quake, or do these need guarded?

**Review by MotoLegacy — CHANGES_REQUESTED** [review]
Thanks for your work on this, left you some feedback and questions.

File: qcsrc/defs.qc
Comment (Zungrysoft) [reply qcsrc/defs.qc:409]: CHAN_NO_PHS_ADD only ever gets used as a modifier to an existing channel. I'll revise the message to make it clear why this will not affect netquake.

File: qcsrc/defs.qc
Comment (Zungrysoft) [reply qcsrc/defs.qc:455]: This is code directly copied from the QuakeWorld codebase. Should we refactor all of defs.qc to use #define statements instead of global constants?

File: qcsrc/doors.qc
Comment (Zungrysoft) [reply qcsrc/doors.qc:66]: Yes. This is straight from the QuakeWorld codebase.

It's kind of weird. It's not actually a channel, it's a flag that modifies how the sound works. From what I understand, this makes it so that the sound packet will be sent to everyone in the server, not just to people within the potentially-hearable-set. (PHS)

File: qcsrc/localized_text.qc
Comment (Zungrysoft) [reply qcsrc/localized_text.qc:68]: Sorry. I meant to fix the screwy indentation in this file when I was done with anything, but I forgot. I'll do that now.

File: qcsrc/localized_text.qc
Comment (MotoLegacy) [reply qcsrc/localized_text.qc]: I dig it, was just curious!

File: qcsrc/defs.qc
Comment (MotoLegacy) [reply qcsrc/defs.qc:455]: Personally I would, just for cleanliness and prog size. Consider it low priority though, I'm fine if you'd want to proceed without it

File: qcsrc/misc.qc
Comment (Zungrysoft) [reply qcsrc/misc.qc:247]: I checked both the default QW progs and NQ progs and they both have 20 for this value. I don't know how it got to be 200 in our progs.

File: qcsrc/doors.qc
Comment (MotoLegacy) [reply qcsrc/doors.qc:66]: Thanks for clearing that up, marking as resolved because one of your previous comments describes a path to document this

File: qcsrc/world.qc
Comment (Zungrysoft) [reply qcsrc/world.qc:260]: Used in deathmatch mode 4 by the backpack mechanic.

File: qcsrc/server.qc
Comment (Zungrysoft) [reply qcsrc/server.qc:1]: It's called server.qc in the original progs, so I figured I would leave it the same. It's kind of a bad name though since it really has nothing to do with server functions. It's just a file for redefining functions that were left out of qw's codebase.

File: qcsrc/spectate.qc
Comment (Zungrysoft) [reply qcsrc/spectate.qc:6]: This is just verbatim from the original progs. What do you think should be changed?

Comment (MotoLegacy) [reply qcsrc/spectate.qc:6]: Seems it was leftover since the purpose line is above it. Can you remove this like to be tidy?

**Comment by lavenderdotpet** [comment]
what changes are currently holding this back

**Comment by Zungrysoft** [comment]
@lavenderdotpet 

The only things I have left to do is determine whether qwprogs.dat should end up in pak0.pak or be unpacked and test a few more things related to deathmatch. Then @MotoLegacy needs to re-review it and answer my followup questions.

**Comment by MotoLegacy** [comment]
Hi @Zungrysoft, I don't see any follow up questions that you asked for me to answer.

**Comment by Zungrysoft** [comment]
@MotoLegacy Actually, I think that's it. Just need your re-review. Thanks!

File: qcsrc/deathmatch_settings.qc
Comment (MotoLegacy) [inline qcsrc/deathmatch_settings.qc:53]: Can we document these infokey changes somewhere? Since it's seemingly off-spec.

Comment (MotoLegacy) [reply qcsrc/deathmatch_settings.qc:53]: Making a note that this is being addressed in #172. Marking as resolved.

**Comment by Zungrysoft** [comment]
@MotoLegacy Great! Thanks! Do I need a second review? Who else in the community has touched the progs.dat code?

**Comment by lavenderdotpet** [comment]
> Thanks for your work on this! Looks good to me.

should i aprove this i dont really know what im looking at when it comes to code but between u and zungry yall are like the main 2 active people who work in code maybe random brushes should review this and not me?

**Comment by MotoLegacy** [comment]
Yes, I will request a review from @RandomBrushes

**Review by RandomBrushes — APPROVED** [review]
Only have one minor comment for client.qc, but that's a stylistic thing regarding nesting if/elseif/else and shall not block merging.

**Pull request merged**
